L1Trigger/L1THGCal/src/backend/HGCalMulticlusteringImpl.cc

```diff
@@ -170,6 +192,314 @@ void HGCalMulticlusteringImpl::clusterizeDBSCAN( const std::vector<edm::Ptr<l1t:
+                int binRight = bin_phi + 1;
+                if( binRight>=int(nBinsPhiHisto_) ) binRight -= nBinsPhiHisto_;
+
+                float MIPT_W = histoClusters[{z_side,bin_R,binLeft}];
+                float MIPT_E = histoClusters[{z_side,bin_R,binRight}];
+                float MIPT_SW = histoClusters[{z_side,bin_R-1,binLeft}];
+                float MIPT_SE = histoClusters[{z_side,bin_R-1,binRight}];
+                float MIPT_NW = histoClusters[{z_side,bin_R+1,binLeft}];
+                float MIPT_NE = histoClusters[{z_side,bin_R+1,binRight}];
+
+                isMax &= MIPT_seed>=MIPT_S;
+                isMax &= MIPT_seed>MIPT_N;
+                isMax &= MIPT_seed>=MIPT_E;
+                isMax &= MIPT_seed>=MIPT_SE;
+                isMax &= MIPT_seed>=MIPT_NE;
+                isMax &= MIPT_seed>MIPT_W;
+                isMax &= MIPT_seed>MIPT_SW;
+                isMax &= MIPT_seed>MIPT_NW;
+
+                if(isMax){
+                    float ROverZ_seed = kROverZMin_ + (bin_R+0.5) * (kROverZMax_-kROverZMin_)/nBinsRHisto_;
+                    float phi_seed = -M_PI + (bin_phi+0.5) * 2*M_PI/nBinsPhiHisto_;
+                    float x_seed = ROverZ_seed*cos(phi_seed);
+                    float y_seed = ROverZ_seed*sin(phi_seed);
+                    seedPositions.emplace_back(x_seed,y_seed,z_side);
+                }
+
+            }
+
+        }
+
+    }
+
+    return seedPositions;
+
+}
+
+
+
+
+
+std::vector<GlobalPoint> HGCalMulticlusteringImpl::computeThresholdSeeds( std::map<std::vector<int>,float> histoClusters ){
+
+    std::vector<GlobalPoint> seedPositions;
+
+    for(int z_side = -1; z_side<2; z_side++){
+        if(z_side==0) continue;
+
+        for(int bin_R = 0; bin_R<int(nBinsRHisto_); bin_R++){
+
+            for(int bin_phi = 0; bin_phi<int(nBinsPhiHisto_); bin_phi++){
+
+                float MIPT_seed = histoClusters[{z_side,bin_R,bin_phi}];
+                bool isSeed = MIPT_seed > histoThreshold_;
+
+                if(isSeed){
+                    float ROverZ_seed = kROverZMin_ + (bin_R+0.5) * (kROverZMax_-kROverZMin_)/nBinsRHisto_;
+                    float phi_seed = -M_PI + (bin_phi+0.5) * 2*M_PI/nBinsPhiHisto_;
+                    float x_seed = ROverZ_seed*cos(phi_seed);
+                    float y_seed = ROverZ_seed*sin(phi_seed);
+                    seedPositions.emplace_back(x_seed,y_seed,z_side);
+                }
+
+            }
+
+        }
+
+    }
+
+    return seedPositions;
+
+}
+
+
+
+std::vector<l1t::HGCalMulticluster> HGCalMulticlusteringImpl::clusterSeedMulticluster(const std::vector<edm::Ptr<l1t::HGCalCluster>> & clustersPtrs,
+										      const std::vector<GlobalPoint> seeds){
+
+
+    std::map<int,l1t::HGCalMulticluster> mapSeedMulticluster;
+    std::vector<l1t::HGCalMulticluster> multiclustersTmp;
+
+    for(std::vector<edm::Ptr<l1t::HGCalCluster>>::const_iterator clu = clustersPtrs.begin(); clu != clustersPtrs.end(); ++clu){
```

Comment:
Would also be better to have a range-based loop here and remove these stacked dereference operators